src/server/generic_family.cc

```diff
@@ -243,161 +251,182 @@ OpResult<RestoreArgs> RestoreArgs::TryFrom(const CmdArgList& args) {
+      src_found_ = KeyExists(t, shard, src_key_);
+      if (src_found_) {
+        return SerializeSrc(t, shard);
+      }
+    }
 
-    res->found = IsValid(it);
-    if (res->found) {
-      res->ref_val = it->second.AsRef();
-      res->expire_ts = db_slice.ExpireTime(exp_it);
-      res->sticky = it->first.IsSticky();
+    if (shard_id == dest_sid_) {
+      dest_found_ = KeyExists(t, shard, dest_key_);
     }
+
     return OpStatus::OK;
   };
 
   t->Execute(std::move(cb), false);
 };
 
-void Renamer::Finalize(Transaction* t, bool skip_exist_dest) {
-  if (!src_res_.found) {
+void Renamer::Finalize(Transaction* t, bool destination_should_not_exist) {
+  if (!src_found_) {
     status_ = OpStatus::KEY_NOTFOUND;
     t->Conclude();
     return;
   }
 
-  if (dest_res_.found && skip_exist_dest) {
+  if (dest_found_ && destination_should_not_exist) {
     status_ = OpStatus::KEY_EXISTS;
     t->Conclude();
     return;
   }
 
-  DCHECK(src_res_.ref_val.IsRef());
-
-  // Src key exist and we need to override the destination.
-  // Alternatively, we could apply an optimistic algorithm and move src at Find step.
-  // We would need to restore the state in case of cleanups.
-  t->Execute([&](Transaction* t, EngineShard* shard) { return MoveSrc(t, shard); }, false);
-  t->Execute([&](Transaction* t, EngineShard* shard) { return UpdateDest(t, shard); }, true);
-}
+  auto cb = [this](Transaction* t, EngineShard* shard) {
+    const ShardId shard_id = shard->shard_id();
 
-OpStatus Renamer::MoveSrc(Transaction* t, EngineShard* es) {
-  if (es->shard_id() == src_sid_) {  // Handle source key.
-    auto res = es->db_slice().FindMutable(t->GetDbContext(), src_res_.key);
-    auto& it = res.it;
-    CHECK(IsValid(it));
-
-    // We distinguish because of the SmallString that is pinned to its thread by design,
-    // thus can not be accessed via another thread.
-    // Therefore, we copy it to standard string in its thread.
-    if (it->second.ObjType() == OBJ_STRING) {
-      it->second.GetString(&str_val_);
-    } else {
-      bool has_expire = it->second.HasExpire();
-      pv_ = std::move(it->second);
-      it->second.SetExpire(has_expire);
+    if (shard_id == src_sid_) {
+      return DelSrc(t, shard);
     }
 
-    res.post_updater.Run();
-    CHECK(es->db_slice().Del(t->GetDbIndex(), it));  // delete the entry with empty value in it.
-    if (es->journal()) {
-      RecordJournal(t->GetOpArgs(es), "DEL", ArgSlice{src_res_.key}, 2);
+    if (shard_id == dest_sid_) {
+      return DeserializeDest(t, shard);
     }
+
+    return OpStatus::OK;
+  };
+
+  t->Execute(std::move(cb), true);
+}
+
+bool Renamer::KeyExists(Transaction* t, EngineShard* shard, std::string_view key) {
+  auto& db_slice = shard->db_slice();
+  auto it = db_slice.FindReadOnly(t->GetDbContext(), key).it;
+  return IsValid(it);
+}
+
+OpStatus Renamer::SerializeSrc(Transaction* t, EngineShard* shard) {
+  auto dump_res = OpDump(t->GetOpArgs(shard), src_key_);
```

Comment:
OpDump calls FindReadOnly and then SerializerBase::DumpObject
than again in line 369 you call FindReadOnly
I think using OpDump is redundant here,
you can call  SerializerBase::DumpObject in this funciton